go/libraries/doltcore/merge/merge_prolly_rows.go

```diff
@@ -1532,40 +1544,131 @@ func (m *valueMerger) tryMerge(left, right, base val.Tuple) (val.Tuple, bool) {
+
+	// We can't just look at the bytes to determine this, because if a cell's byte representation changed,
+	// but only because of a schema change, we shouldn't consider that a conflict.
+	// Conversely, if there was a schema change on only one side, we shouldn't consider the cells equal
+	// even if they have the same bytes.
+
+	// Thus, we must convert all cells to the type in the result schema before comparing them.
+
+	if baseCol != nil {
+		var ok bool
+		baseCol, ok = convert(ctx, m.baseVD, m.resultVD, m.resultSchema, baseColIdx, i, base, baseCol, m.ns)
+		if !ok {
+			return nil, true
+		}
 	}
 
-	if base == nil {
-		// Conflicting insert
-		return nil, true
+	var leftModified, rightModified bool
+
+	if leftColIdx == -1 && rightColIdx == -1 {
+		// Both branches are implicitly NULL
+		return nil, false
 	}
 
-	var baseVal []byte
-	if b := m.baseMapping[i]; b != -1 {
-		baseVal = base.GetField(b)
+	if rightColIdx == -1 {
+		// The right branch is implicitly NULL
+		rightModified = baseCol != nil
+	} else {
+		// Attempt to convert the right column to match the result schema, then compare it to the base.
+		var ok bool
+		rightCol, ok = convert(ctx, m.rightVD, m.resultVD, m.resultSchema, rightColIdx, i, right, rightCol, m.ns)
+		if !ok {
+			return nil, true
+		}
+		rightModified = m.resultVD.Comparator().CompareValues(i, rightCol, baseCol, resultType) != 0
+	}
+
+	// The left value was previously converted in `migrateDataToMergedSchema`, so we don't need to convert it here.
+	if m.resultVD.Comparator().CompareValues(i, leftCol, rightCol, resultType) == 0 {
+		// columns are equal, return either.
+		return leftCol, false
 	}
 
-	leftModified := m.vD.Comparator().CompareValues(i, leftCol, baseVal, m.vD.Types[i]) != 0
-	rightModified := m.vD.Comparator().CompareValues(i, rightCol, baseVal, m.vD.Types[i]) != 0
+	leftModified = m.resultVD.Comparator().CompareValues(i, leftCol, baseCol, resultType) != 0
 
 	switch {
 	case leftModified && rightModified:
+		// concurrent modification
 		return nil, true
 	case leftModified:
 		return leftCol, false
 	default:
 		return rightCol, false
 	}
 }
+
+func getColumn(tuple *val.Tuple, mapping *val.OrdinalMapping, idx int) (col []byte, colIndex int, exists bool) {
+	colIdx := (*mapping)[idx]
+	if colIdx == -1 {
+		return nil, -1, false
+	}
+	return tuple.GetField(colIdx), colIdx, true
+}
+
+// convert takes the `i`th column in the provided tuple and converts it to the type specified in the provided schema.
+// returns the new representation, and a bool indicating success.
+func convert(ctx context.Context, fromDesc, toDesc val.TupleDesc, toSchema schema.Schema, fromIndex, toIndex int, tuple val.Tuple, originalValue []byte, ns tree.NodeStore) ([]byte, bool) {
+	if fromDesc.Types[fromIndex] == toDesc.Types[toIndex] {
+		// No conversion is necessary here.
+		return originalValue, true
+	}
+	parsedCell, err := index.GetField(ctx, fromDesc, fromIndex, tuple, ns)
+	if err != nil {
+		return nil, true
```

Comment:
Looks like this should be returning `false` instead of `true`

It's probably also worth doing something with that `err`. If we're trying to debug a customer issue in this code, having that err logged somewhere could be handy. One idea is grabbing the server log from `ctx` and logging it as a warning or error.

Reply:
Similar comment for the other `err` on line 1667. Seems like the only way to see that currently would be to run the code through the debugger, but it would probably be helpful to have that in the logs or somewhere a customer could easily report to us.

Reply:
Good point. I think that we should log the error somehow and then return that the conversion failed, since it likely indicates an unforeseen schema conflict preventing us from converting to the new column type.

I couldn't find anything about a server log, can you point me to it?

Reply:
On second thought, I'm just going to propagate the error.

It's possible that as we allow more type changes in schema merges, some of these errors should be treated as data conflicts and shouldn't be propagated. But we can treat those on a case-by-case basis by inspecting the type of the error. For now, always propagating the error ensures that we're not suppressing important errors that *aren't* do to type conversion conflicts.